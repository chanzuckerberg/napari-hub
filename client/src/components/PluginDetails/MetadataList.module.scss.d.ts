export type Styles = {
  commaList: string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
