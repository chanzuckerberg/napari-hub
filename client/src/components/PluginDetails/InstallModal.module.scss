// Make SVG stroke color black
.closeIcon {
  path {
    @apply stroke-current text-black;
  }
}
