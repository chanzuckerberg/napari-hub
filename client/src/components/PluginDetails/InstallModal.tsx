import clsx from 'clsx';
import Image from 'next/image';
import { useRef } from 'react';
import { useClickAway } from 'react-use';

import { Overlay } from '@/components/common';
import { Fade } from '@/components/common/animations';
import { Close, Copy } from '@/components/common/icons';
import { MediaFragment } from '@/components/common/media';
import { usePluginState } from '@/context/plugin';

import styles from './InstallModal.module.scss';

/**
 * Component that renders an inline button for copying the plugin name to the
 * user's clipboard.
 */
function CopyPluginNameButton() {
  const { plugin } = usePluginState();

  return (
    <button
      className={clsx(
        // Button colors
        'bg-napari-light hover:bg-napari-primary',

        // Show button inline with text.
        'inline',

        // Padding
        'py-px px-1',

        // Animate colors when hovering over button
        'transition-colors',
      )}
      onClick={() => navigator.clipboard.writeText(plugin.name)}
      type="button"
    >
      {plugin.name} <Copy className="inline" />
    </button>
  );
}

interface Closeable {
  /**
   * Callback to close the modal when the user clicks away or clicks the close
   * buttons.
   */
  onClose(): void;
}

/**
 * Component that renders the modal body.  This includes the instructions on
 * how to install the napari plugin and a button to copy the plugin name.
 */
function InstallModalBody({ onClose }: Closeable) {
  return (
    <>
      {/* Header showing title and close button */}
      <header className="flex justify-between mb-9">
        <h2 className="font-bold text-2xl">Installing a plugin with napari</h2>

        {/* Close button */}
        <MediaFragment greaterThanOrEqual="sm">
          <button onClick={onClose} type="button">
            <Close className={styles.closeIcon} />
          </button>
        </MediaFragment>
      </header>

      {/* Numbered list of instructions for installing a plugin */}
      <ol className="list-decimal list-inside font-bold">
        <li>
          <p className="font-normal inline leading-8">
            From the “Plugins” menu within the napari application, select
            “Install/Uninstall Package(s)...”.
          </p>

          <div className="my-3">
            <Image
              src="/images/plugin-install-menu.png"
              alt="napari plugin install menu"
              width={141}
              height={74}
            />
          </div>
        </li>

        <li>
          <p className="font-normal inline leading-8">
            Copy <CopyPluginNameButton /> and paste it where it says “Install by
            name/url…”
          </p>

          <div className="my-3">
            <Image
              src="/images/plugin-list.png"
              alt="napari plugin list"
              width={430}
              height={191}
            />
          </div>
        </li>

        <li>
          <p className="font-normal inline">Click “Install”.</p>
          <p className="font-normal italic text-xs my-6">
            To get started with napari, visit{' '}
            <a
              className="underline hover:text-napari-primary"
              href="https://napari.org"
              target="_blank"
              rel="noreferrer"
            >
              napari.org
            </a>
            .
          </p>
        </li>
      </ol>
    </>
  );
}

interface InstallModalProps extends Closeable {
  /**
   * Determines if the modal should be visible or not.
   */
  visible: boolean;
}

/**
 * Component for rendering the installation modal.  This component handles
 * animating the modal into / out of view, displaying an overlay when the modal
 * is visible, and closing the modal when the user clicks away from it.
 */
export function InstallModal({ onClose, visible }: InstallModalProps) {
  const modalRef = useRef<HTMLDivElement | null>(null);
  useClickAway(modalRef, onClose);

  return (
    <>
      <Overlay visible={visible} />

      <div className="absolute min-w-napari-xs">
        <Fade
          className={clsx(
            // White background and drop shadow
            'bg-white shadow-2xl',

            // Add scrollbar if modal contents are too long.
            'overflow-y-auto',

            // Render modal fixed to screen and above everything else
            'fixed z-50',

            // Padding
            'p-6 md:p-12',

            // Dimensions
            'w-5/6 max-w-[775px] max-h-[706px]',

            // Positioning: This centers the modal in the middle of the viewport
            'top-1/2 left-1/2',
            'transform -translate-x-1/2 -translate-y-1/2',
          )}
          ref={modalRef}
          visible={visible}
        >
          <InstallModalBody onClose={onClose} />

          <div className="flex justify-end">
            <button
              className="border-2 py-4 px-6 border-napari-primary"
              onClick={onClose}
              type="button"
            >
              Close
            </button>
          </div>
        </Fade>
      </div>
    </>
  );
}
