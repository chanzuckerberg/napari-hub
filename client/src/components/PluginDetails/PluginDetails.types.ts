/**
 * Metadata item to render in the MetadataList component.
 */
export interface MetadataItem {
  title: string;
  value: string | string[];
}
