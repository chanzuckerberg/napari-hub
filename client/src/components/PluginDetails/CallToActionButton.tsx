import clsx from 'clsx';
import { useState } from 'react';

import { InstallModal } from './InstallModal';

interface Props {
  className?: string;
}

/**
 * Component that renders the CTA button and installation modal.  Clicking the
 * install button will render the modal visible.
 */
export function CallToActionButton({ className }: Props) {
  const [visible, setVisible] = useState(false);

  return (
    <>
      <InstallModal onClose={() => setVisible(false)} visible={visible} />

      <button
        className={clsx(
          className,

          // Button color
          'bg-napari-primary',

          // Dimensions
          'h-12 w-full lg:max-w-napari-col',
        )}
        onClick={() => setVisible(true)}
        type="button"
      >
        Install
      </button>
    </>
  );
}
