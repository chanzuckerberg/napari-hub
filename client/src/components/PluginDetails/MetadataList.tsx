import clsx from 'clsx';
import { ReactNode } from 'react-markdown';

import styles from './MetadataList.module.scss';
import { MetadataItem, MetadataValueTypes } from './PluginDetails.types';

interface CommonProps {
  /**
   * Renders the metadata list with headings and values inline.
   */
  inline?: boolean;
}

interface MetadataListItemProps extends CommonProps {
  /**
   * Title of the current metadata.
   */
  title: string;

  /**
   * List of values for the current metadata.
   */
  values: MetadataValueTypes[];
}

/**
 * Component for rendering a metadata list item.  This renders the title
 * heading and metadata value.
 */
function MetadataListItem({ inline, title, values }: MetadataListItemProps) {
  const isEmpty = values.filter(Boolean).length === 0;

  return (
    <li className="mb-4 text-sm">
      <h4
        className={clsx(
          // Inline styles
          inline ? 'inline mr-2' : 'block',

          // Font
          'font-bold whitespace-nowrap',
        )}
      >
        {title}:
      </h4>

      {/* Render placeholder for empty data.  */}
      {isEmpty && (
        <p
          className={clsx(
            'text-gray-400 font-normal',
            inline ? 'inline' : 'block',
          )}
        >
          information not submitted
        </p>
      )}

      {!isEmpty && (
        // Render metadata values as list
        <ul className={clsx('list-none', inline ? 'inline' : 'block')}>
          {values.map((value) => {
            let node: ReactNode;
            let key: string;

            if (typeof value === 'string') {
              key = value;
              node = value;
            } else {
              // If metadata value is link, render icon and anchor tag.
              key = `${value.text}-${value.href}`;
              node = (
                <span
                  className={clsx(
                    inline ? 'inline-flex' : 'flex',
                    'items-center',
                  )}
                >
                  {value.icon}

                  <a
                    className="ml-2 underline hover:text-napari-primary"
                    href={value.href}
                    target="_blank"
                    rel="noreferrer"
                  >
                    {value.text}
                  </a>
                </span>
              );
            }

            return (
              <li
                className={clsx(
                  // Margins
                  'my-2 last:mb-0',

                  // Line height
                  'leading-8',

                  // Render as comma list if inline
                  // https://markheath.net/post/css-comma-separated-list
                  inline && ['inline', styles.commaList],
                )}
                key={key}
              >
                {node}
              </li>
            );
          })}
        </ul>
      )}
    </li>
  );
}

interface MetadataListProps extends CommonProps {
  /**
   * Class name to pass to root component.
   */
  className?: string;

  /**
   * Renders the metadata list horizontally.
   */
  horizontal?: boolean;

  /**
   * List of values for the current metadata.
   */
  items: MetadataItem[];
}

/**
 * Component for rendering a list of plugin metadata titles and values.
 */
export function MetadataList({
  className,
  horizontal,
  inline,
  items,
}: MetadataListProps) {
  return (
    <ul
      className={clsx(
        'list-none',
        className,
        horizontal && 'grid grid-cols-napari-3-col-fr',
      )}
    >
      {items.map((item) => {
        const values = item.value instanceof Array ? item.value : [item.value];

        return (
          <MetadataListItem
            key={`${item.title}-${String(values)}`}
            inline={inline}
            title={item.title}
            values={values}
          />
        );
      })}
    </ul>
  );
}
