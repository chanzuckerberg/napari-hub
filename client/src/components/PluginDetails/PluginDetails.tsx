import clsx from 'clsx';

import { ColumnLayout, Markdown } from '@/components/common';
import { Media, MediaFragment } from '@/components/common/media';
import { usePluginState } from '@/context/plugin';

import { CallToActionButton } from './CallToActionButton';
import { PluginMetadata } from './PluginMetadata';
import { SupportInfo } from './SupportInfo';

function PluginLeftColumn() {
  return (
    <Media greaterThanOrEqual="3xl">
      <PluginMetadata />
    </Media>
  );
}

function PluginCenterColumn() {
  const { plugin } = usePluginState();

  return (
    <article className="w-full col-span-2 xl:col-span-3">
      <h1 className="font-bold text-4xl">{plugin.name}</h1>
      <h2 className="font-semibold my-6 text-lg">{plugin.summary}</h2>

      <Media
        className={clsx(
          // Layout
          'flex flex-col',

          // Align CTA and metadata link horizontally for lg layouts
          'lg:flex-row lg:items-center',

          // Margins
          'my-6 md:my-12',
        )}
        lessThan="3xl"
      >
        <MediaFragment lessThan="2xl">
          <CallToActionButton />
        </MediaFragment>

        <a
          className={clsx(
            // Text styling
            'underline hover:text-napari-primary',

            /*
              Top margins: This is used for smaller layouts because the CTA
              button is above the metadata link.
            */
            'mt-6 md:mt-12 lg:mt-0',

            /*
              Left margins: This is used when the CTA and metadata link are
              inline.  The margin is removed when the CTA moves to the right
              column on 2xl layouts.
            */
            'lg:ml-12 2xl:ml-0',
          )}
          href="#pluginMetadata"
        >
          View project data
        </a>
      </Media>

      <SupportInfo className="mb-6 md:mb-12" />

      <Markdown className="mb-10" disableHeader>
        {plugin.description}
      </Markdown>

      <CallToActionButton className="mb-6 md:mb-12 2xl:mb-20" />

      <MediaFragment lessThan="3xl">
        <PluginMetadata />
      </MediaFragment>
    </article>
  );
}

function PluginRightColumn() {
  const { plugin } = usePluginState();

  return (
    <Media greaterThanOrEqual="2xl">
      {/*  Keep button on screen when scrolling on 2xl. */}
      <CallToActionButton className="fixed" />

      <Markdown.TOC
        className="fixed flex mt-24"
        markdown={plugin.description}
      />
    </Media>
  );
}

/**
 * Component for rendering the plugin details page.
 */
export function PluginDetails() {
  return (
    <ColumnLayout className="p-6 md:p-12 2xl:px-0" data-testid="pluginDetails">
      <PluginLeftColumn />
      <PluginCenterColumn />
      <PluginRightColumn />
    </ColumnLayout>
  );
}
