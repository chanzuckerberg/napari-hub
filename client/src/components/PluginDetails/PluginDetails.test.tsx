import { render, screen } from '@testing-library/react';

import { PluginStateProvider } from '@/context/plugin';
import napariPlugin from '@/fixtures/napari.json';

import { PluginDetails } from './PluginDetails';

describe('<PluginDetails />', () => {
  it('should match snapshot', () => {
    render(
      <PluginStateProvider plugin={napariPlugin}>
        <PluginDetails />
      </PluginStateProvider>,
    );
    expect(screen.getByTestId('pluginDetails')).toMatchSnapshot();
  });
});
