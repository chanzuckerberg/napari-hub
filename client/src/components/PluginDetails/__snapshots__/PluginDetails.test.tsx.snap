// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PluginDetails /> should match snapshot 1`] = `
<div
  class="flex 2xl:grid 2xl:justify-center p-6 md:p-14 2xl:px-0 2xl:gap-14 2xl:grid-cols-napari-2-col 3xl:grid-cols-napari-3-col"
  data-testid="pluginDetails"
>
  <div
    class="fresnel-container fresnel-greaterThanOrEqual-3xl "
  >
    <div
      class="flex flex-col xl:grid xl:grid-cols-[1fr,1fr,1fr] 3xl:flex"
      id="pluginMetadata"
    >
      <ul
        class="list-none"
      >
        <li
          class="mb-4 text-sm"
        >
          <h4
            class="inline 3xl:block font-bold whitespace-nowrap mr-2 3xl:m-0"
          >
            Version
            :
          </h4>
          <ul
            class="fresnel-greaterThanOrEqual-3xl list-none"
          >
            <li
              class="my-2 last:mb-0"
            >
              0.0.2
            </li>
          </ul>
        </li>
        <li
          class="mb-4 text-sm"
        >
          <h4
            class="inline 3xl:block font-bold whitespace-nowrap mr-2 3xl:m-0"
          >
            Release date
            :
          </h4>
          <ul
            class="fresnel-greaterThanOrEqual-3xl list-none"
          >
            <li
              class="my-2 last:mb-0"
            >
              03 March 2021
            </li>
          </ul>
        </li>
        <li
          class="mb-4 text-sm"
        >
          <h4
            class="inline 3xl:block font-bold whitespace-nowrap mr-2 3xl:m-0"
          >
            First released
            :
          </h4>
          <ul
            class="fresnel-greaterThanOrEqual-3xl list-none"
          >
            <li
              class="my-2 last:mb-0"
            >
              22 February 2021
            </li>
          </ul>
        </li>
        <li
          class="mb-4 text-sm"
        >
          <h4
            class="inline 3xl:block font-bold whitespace-nowrap mr-2 3xl:m-0"
          >
            Development status
            :
          </h4>
          <ul
            class="fresnel-greaterThanOrEqual-3xl list-none"
          >
            <li
              class="my-2 last:mb-0"
            >
              4 - Beta
            </li>
          </ul>
        </li>
        <li
          class="mb-4 text-sm"
        >
          <h4
            class="inline 3xl:block font-bold whitespace-nowrap mr-2 3xl:m-0"
          >
            License
            :
          </h4>
          <ul
            class="fresnel-greaterThanOrEqual-3xl list-none"
          >
            <li
              class="my-2 last:mb-0"
            >
              MIT
            </li>
          </ul>
        </li>
      </ul>
      <div
        class="fresnel-greaterThanOrEqual-3xl my-6 h-px bg-black border-none"
      />
      <div
        class="flex flex-col items-start xl:items-center 2xl:items-start text-sm"
      >
        <h4
          class="font-bold"
        >
          Github Activity
        </h4>
        <ul
          class="list-none"
        >
          <li
            class="my-2"
          >
            Stars
            : 
            <span
              class="font-bold"
            >
              0
            </span>
          </li>
          <li
            class="my-2"
          >
            Forks
            : 
            <span
              class="font-bold"
            >
              0
            </span>
          </li>
          <li
            class="my-2"
          >
            Issues + PRs
            : 
            <span
              class="font-bold"
            >
              0
            </span>
          </li>
        </ul>
      </div>
      <div
        class="fresnel-greaterThanOrEqual-3xl my-6 h-px bg-black border-none"
      />
      <ul
        class="list-none"
      >
        <li
          class="mb-4 text-sm"
        >
          <h4
            class="inline 3xl:block font-bold whitespace-nowrap mr-2 3xl:m-0"
          >
            Python versions supported
            :
          </h4>
          <ul
            class="fresnel-greaterThanOrEqual-3xl list-none"
          >
            <li
              class="my-2 last:mb-0"
            >
              &gt;=3.6
            </li>
          </ul>
        </li>
        <li
          class="mb-4 text-sm"
        >
          <h4
            class="inline 3xl:block font-bold whitespace-nowrap mr-2 3xl:m-0"
          >
            Operating system
            :
          </h4>
          <ul
            class="fresnel-greaterThanOrEqual-3xl list-none"
          >
            <li
              class="my-2 last:mb-0"
            >
              OS Independent
            </li>
          </ul>
        </li>
        <li
          class="mb-4 text-sm"
        >
          <h4
            class="inline 3xl:block font-bold whitespace-nowrap mr-2 3xl:m-0"
          >
            Requirements
            :
          </h4>
          <ul
            class="fresnel-greaterThanOrEqual-3xl list-none"
          >
            <li
              class="my-2 last:mb-0"
            >
              napari-plugin-engine (&gt;=0.1.4)
            </li>
            <li
              class="my-2 last:mb-0"
            >
              numpy
            </li>
            <li
              class="my-2 last:mb-0"
            >
              zarr
            </li>
            <li
              class="my-2 last:mb-0"
            >
              dask[complete]
            </li>
          </ul>
        </li>
      </ul>
    </div>
  </div>
  <article
    class="w-full"
  >
    <h1
      class="font-bold text-4xl"
    >
      napari-compressed-labels-io
    </h1>
    <h2
      class="font-semibold my-6 text-lg"
    >
      Plugin exploring different options for reading and writing compressed and portable labels layers in napari.
    </h2>
    <div
      class="fresnel-container fresnel-lessThan-3xl flex flex-col lg:flex-row lg:items-center my-6 md:my-12"
    >
      <div
        class="absolute min-w-napari-xs"
      />
      <button
        class="fresnel-lessThan-2xl bg-napari-primary 2xl:fixed h-12 w-full lg:max-w-napari-side-col"
        type="button"
      >
        Install
      </button>
      <a
        class="underline hover:text-napari-primary mt-6 md:mt-12 lg:mt-0 lg:ml-12 2xl:ml-0"
        href="#pluginMetadata"
      >
        View project data
      </a>
    </div>
    <div
      class="mb-10 markdown prose max-w-none"
    >
      

      <p>
        <a
          href="https://github.com/DragaDoncila/napari-compressed-labels-io/raw/master/LICENSE"
        >
          <img
            alt="License"
            src="https://img.shields.io/pypi/l/napari-compressed-labels-io.svg?color=green"
          />
        </a>
        

        <a
          href="https://pypi.org/project/napari-compressed-labels-io"
        >
          <img
            alt="PyPI"
            src="https://img.shields.io/pypi/v/napari-compressed-labels-io.svg?color=green"
          />
        </a>
        

        <a
          href="https://python.org"
        >
          <img
            alt="Python Version"
            src="https://img.shields.io/pypi/pyversions/napari-compressed-labels-io.svg?color=green"
          />
        </a>
        

        <a
          href="https://github.com/DragaDoncila/napari-compressed-labels-io/actions"
        >
          <img
            alt="tests"
            src="https://github.com/DragaDoncila/napari-compressed-labels-io/workflows/tests/badge.svg"
          />
        </a>
        

        <a
          href="https://codecov.io/gh/DragaDoncila/napari-compressed-labels-io"
        >
          <img
            alt="codecov"
            src="https://codecov.io/gh/DragaDoncila/napari-compressed-labels-io/branch/master/graph/badge.svg"
          />
        </a>
      </p>
      

      <h2
        id="description"
      >
        Description
      </h2>
      

      <p>
        This napari plugin provides readers and writers for labels and their corresponding image layers into zarr format for compression and portability. Each reader/writer pair supports a round trip of saving and loading image and labels layers.
      </p>
      

      <h2
        id="writers"
      >
        Writers
      </h2>
      

      <p>
        Two writers are provided by this plugin, each with its own reader.
      </p>
      

      <h3
        id="labels_to_zarr"
      >
        <code
          node="[object Object]"
        >
          labels_to_zarr
        </code>
      </h3>
      

      <p>
        This writer is an alternative to napari's default label writer and will write an entire labels layer, regardless of its dimensions, into a single zarr file. This writer provides the best compression option and its associated reader 
        <code
          node="[object Object]"
        >
          get_zarr_labels
        </code>
         will read the layer back into napari.
      </p>
      

      <p>
        This writer will be called when the user tries to save a selected labels layer into a path ending with .zarr
      </p>
      

      <h3
        id="label_image_pairs_to_zarr"
      >
        <code
          node="[object Object]"
        >
          label_image_pairs_to_zarr
        </code>
      </h3>
      

      <p>
        This writer will save 3-dimensional labels and image layers from the viewer into individual zarrs for portability and convenience. For example, given one labels and one image layer of the shape (10, 200, 200) saved to my_stacks.zarr, 10 subdirectories will be created, each with two zarrs inside of shape (200, 200) corresponding to the labels and image layer.
      </p>
      

      <p>
        This writer allows users to load stacks of associated images, label them, and then quickly save these stacks out into individual slices for easy loading, viewing and interaction. Its associated reader supports the loading into napari of the whole stack, all layers at one slice of the stack, and an individual layer of a given slice of the stack.
      </p>
      

      <p>
        The writer currently supports only 3D layers, with the exception of RGB images of the form (z, y, x, 3), which are also supported.
      </p>
      

      <h2
        id="readers"
      >
        Readers
      </h2>
      

      <p>
        Two readers are provided by this plugin for loading the formats saved by each writer. These are detailed below.
      </p>
      

      <h3
        id="get_zarr_labels"
      >
        <code
          node="[object Object]"
        >
          get_zarr_labels
        </code>
      </h3>
      

      <p>
        This reader will open any zarr file with a .zarray at the top level in 
        <code
          node="[object Object]"
        >
          path
        </code>
         as a labels layer. This is to be used in conjunction with 
        <code
          node="[object Object]"
        >
          labels_to_zarr
        </code>
        .
      </p>
      

      <h3
        id="get_label_image_stack"
      >
        <code
          node="[object Object]"
        >
          get_label_image_stack
        </code>
      </h3>
      

      <p>
        This reader will open any zarr containing a 
        <code
          node="[object Object]"
        >
          .zmeta
        </code>
         file as layers into napari. Depending on what is being opened, the reader will either load a full stack of labels and images, one slice of a stack of images and labels or an individual layer within a slice. This is to be used in conjunction with 
        <code
          node="[object Object]"
        >
          label_image_pairs_to_zarr
        </code>
        .
      </p>
      

      <h2
        id="zmeta"
      >
        .zmeta
      </h2>
      

      <p>
        This metadata file contains information about the layer types in the stack and in each individual slice, as well as the number of image/label slices. This allows the reader plugin to load the correct layer types with appropriate names both at a stack level and at the individual slice level.
      </p>
      

      <h3
        id="an-example-zmeta-specification"
      >
        An example .zmeta specification
      </h3>
      

      <pre>
        <code
          language="json"
          node="[object Object]"
        >
          {
    "meta": {
        "stack": 7                               # number of slices in the entire stack (1 for an individual slice, 0 for a layer within a slice)
    },
    "data": {
        "image" : [                              # all image layers must be listed here
            {
                "name": "leaves_example_data",
                "shape": [790, 790, 3],
                "dtype": "uint8",
                "rgb": true                      # where rgb is false the image will be loaded as greyscale (colormap support has not yet been implemented)
            }
        ],
        "labels" : [
            {
                "name": "oak",
                "shape": [790, 790],
                "dtype": "int64"
            },
            {
                "name": "bg",
                "shape": [790, 790],
                "dtype": "int64"
            }
        ]
    }
}

        </code>
      </pre>
      

      <hr />
      

      <p>
        This 
        <a
          href="https://github.com/napari/napari"
        >
          napari
        </a>
         plugin was generated with 
        <a
          href="https://github.com/audreyr/cookiecutter"
        >
          Cookiecutter
        </a>
         using with 
        <a
          href="https://github.com/napari"
        >
          @napari
        </a>
        's 
        <a
          href="https://github.com/napari/cookiecutter-napari-plugin"
        >
          cookiecutter-napari-plugin
        </a>
         template.
      </p>
      

      <h2
        id="installation"
      >
        Installation
      </h2>
      

      <p>
        You can install 
        <code
          node="[object Object]"
        >
          napari-compressed-labels-io
        </code>
         via 
        <a
          href="https://pypi.org/project/pip/"
        >
          pip
        </a>
        :
      </p>
      

      <pre>
        <code
          language=""
          node="[object Object]"
        >
          pip install napari-compressed-labels-io
        </code>
      </pre>
      

      <h2
        id="contributing"
      >
        Contributing
      </h2>
      

      <p>
        Contributions are very welcome. Tests can be run with 
        <a
          href="https://tox.readthedocs.io/en/latest/"
        >
          tox
        </a>
        , please ensure
the coverage at least stays the same before you submit a pull request.
      </p>
      

      <h2
        id="license"
      >
        License
      </h2>
      

      <p>
        Distributed under the terms of the 
        <a
          href="http://opensource.org/licenses/MIT"
        >
          MIT
        </a>
         license,
"napari-compressed-labels-io" is free and open source software
      </p>
      

      <h2
        id="issues"
      >
        Issues
      </h2>
      

      <p>
        If you encounter any problems, please 
        <a
          href="https://github.com/DragaDoncila/napari-compressed-labels-io/issues"
        >
          file an issue
        </a>
         along with a detailed description.
      </p>
    </div>
    <div
      class="fresnel-lessThan-3xl flex flex-col xl:grid xl:grid-cols-[1fr,1fr,1fr] 3xl:flex"
      id="pluginMetadata"
    >
      <ul
        class="list-none"
      >
        <li
          class="mb-4 text-sm"
        >
          <h4
            class="inline 3xl:block font-bold whitespace-nowrap mr-2 3xl:m-0"
          >
            Version
            :
          </h4>
          <p
            class="fresnel-lessThan-3xl inline"
          >
            0.0.2
          </p>
        </li>
        <li
          class="mb-4 text-sm"
        >
          <h4
            class="inline 3xl:block font-bold whitespace-nowrap mr-2 3xl:m-0"
          >
            Release date
            :
          </h4>
          <p
            class="fresnel-lessThan-3xl inline"
          >
            03 March 2021
          </p>
        </li>
        <li
          class="mb-4 text-sm"
        >
          <h4
            class="inline 3xl:block font-bold whitespace-nowrap mr-2 3xl:m-0"
          >
            First released
            :
          </h4>
          <p
            class="fresnel-lessThan-3xl inline"
          >
            22 February 2021
          </p>
        </li>
        <li
          class="mb-4 text-sm"
        >
          <h4
            class="inline 3xl:block font-bold whitespace-nowrap mr-2 3xl:m-0"
          >
            Development status
            :
          </h4>
          <p
            class="fresnel-lessThan-3xl inline"
          >
            4 - Beta
          </p>
        </li>
        <li
          class="mb-4 text-sm"
        >
          <h4
            class="inline 3xl:block font-bold whitespace-nowrap mr-2 3xl:m-0"
          >
            License
            :
          </h4>
          <p
            class="fresnel-lessThan-3xl inline"
          >
            MIT
          </p>
        </li>
      </ul>
      <div
        class="fresnel-lessThan-xl my-6 h-px bg-black border-none"
      />
      <div
        class="flex flex-col items-start xl:items-center 2xl:items-start text-sm"
      >
        <h4
          class="font-bold"
        >
          Github Activity
        </h4>
        <ul
          class="list-none"
        >
          <li
            class="my-2"
          >
            Stars
            : 
            <span
              class="font-bold"
            >
              0
            </span>
          </li>
          <li
            class="my-2"
          >
            Forks
            : 
            <span
              class="font-bold"
            >
              0
            </span>
          </li>
          <li
            class="my-2"
          >
            Issues + PRs
            : 
            <span
              class="font-bold"
            >
              0
            </span>
          </li>
        </ul>
      </div>
      <div
        class="fresnel-lessThan-xl my-6 h-px bg-black border-none"
      />
      <ul
        class="list-none"
      >
        <li
          class="mb-4 text-sm"
        >
          <h4
            class="inline 3xl:block font-bold whitespace-nowrap mr-2 3xl:m-0"
          >
            Python versions supported
            :
          </h4>
          <p
            class="fresnel-lessThan-3xl inline"
          >
            &gt;=3.6
          </p>
        </li>
        <li
          class="mb-4 text-sm"
        >
          <h4
            class="inline 3xl:block font-bold whitespace-nowrap mr-2 3xl:m-0"
          >
            Operating system
            :
          </h4>
          <p
            class="fresnel-lessThan-3xl inline"
          >
            OS Independent
          </p>
        </li>
        <li
          class="mb-4 text-sm"
        >
          <h4
            class="inline 3xl:block font-bold whitespace-nowrap mr-2 3xl:m-0"
          >
            Requirements
            :
          </h4>
          <p
            class="fresnel-lessThan-3xl inline"
          >
            napari-plugin-engine (&gt;=0.1.4), numpy, zarr, dask[complete]
          </p>
        </li>
      </ul>
    </div>
  </article>
  <div
    class="fresnel-container fresnel-greaterThanOrEqual-2xl "
  >
    <div
      class="absolute min-w-napari-xs"
    />
    <button
      class="bg-napari-primary 2xl:fixed h-12 w-full lg:max-w-napari-side-col"
      type="button"
    >
      Install
    </button>
    <ul
      class="fixed flex mt-24 flex flex-col border-l border-black"
    >
      <li
        class="flex pl-6 h-6 border-l-4 my-2 first:mt-0 last:mb-0 transition-colors hover:border-napari-primary border-transparent"
        data-active="false"
        data-testid="tocItem"
      >
        <a
          class=""
          href="#description"
        >
          Description
        </a>
      </li>
      <li
        class="flex pl-6 h-6 border-l-4 my-2 first:mt-0 last:mb-0 transition-colors hover:border-napari-primary border-transparent"
        data-active="false"
        data-testid="tocItem"
      >
        <a
          class=""
          href="#writers"
        >
          Writers
        </a>
      </li>
      <li
        class="flex pl-6 h-6 border-l-4 my-2 first:mt-0 last:mb-0 transition-colors hover:border-napari-primary border-transparent"
        data-active="false"
        data-testid="tocItem"
      >
        <a
          class=""
          href="#readers"
        >
          Readers
        </a>
      </li>
      <li
        class="flex pl-6 h-6 border-l-4 my-2 first:mt-0 last:mb-0 transition-colors hover:border-napari-primary border-transparent"
        data-active="false"
        data-testid="tocItem"
      >
        <a
          class=""
          href="#zmeta"
        >
          .zmeta
        </a>
      </li>
      <li
        class="flex pl-6 h-6 border-l-4 my-2 first:mt-0 last:mb-0 transition-colors hover:border-napari-primary border-transparent"
        data-active="false"
        data-testid="tocItem"
      >
        <a
          class=""
          href="#installation"
        >
          Installation
        </a>
      </li>
      <li
        class="flex pl-6 h-6 border-l-4 my-2 first:mt-0 last:mb-0 transition-colors hover:border-napari-primary border-transparent"
        data-active="false"
        data-testid="tocItem"
      >
        <a
          class=""
          href="#contributing"
        >
          Contributing
        </a>
      </li>
      <li
        class="flex pl-6 h-6 border-l-4 my-2 first:mt-0 last:mb-0 transition-colors hover:border-napari-primary border-transparent"
        data-active="false"
        data-testid="tocItem"
      >
        <a
          class=""
          href="#license"
        >
          License
        </a>
      </li>
      <li
        class="flex pl-6 h-6 border-l-4 my-2 first:mt-0 last:mb-0 transition-colors hover:border-napari-primary border-black"
        data-active="true"
        data-testid="tocItem"
      >
        <a
          class="font-bold"
          href="#issues"
        >
          Issues
        </a>
      </li>
    </ul>
  </div>
</div>
`;
